src/openms/source/FILTERING/DATAREDUCTION/MassTraceDetection.cpp

```diff
@@ -333,10 +426,10 @@ namespace OpenMS
         updateIterativeWeightedMeanMZ(apex_peak.getMZ(), apex_peak.getIntensity(), centroid_mz, prev_counter, prev_denom);
 
         std::vector<std::pair<Size, Size> > gathered_idx;
-        gathered_idx.emplace_back(apex_scan_idx, apex_peak_idx);
-        if (fwhm_meta_idx != -1)
+        gathered_idx.emplace_back(chrom_apices[index].scan_idx_,chrom_apices[index].peak_idx_);
+        if (fwhm_meta_idx)
         {
-          fwhms_mz.push_back(work_exp[apex_scan_idx].getFloatDataArrays()[fwhm_meta_idx][apex_peak_idx]);
+          fwhms_mz.push_back(work_exp[chrom_apices[index].scan_idx_].getFloatDataArrays()[0][chrom_apices[index].peak_idx_]);
```

Comment:
this logic here is a bit flawed.
There is no guarantee that the meta information is at index '0'.
This was not correctly implemented in the previous version IMHO, but your refactoring makes it harder to determine the correct index. Please revert that part.